File: 300.600/352.DataStreamAsDisjointIntervals.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <algorithm>
using namespace std;

class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
        data.clear();
    }

    void addNum(int val) {
        auto it = lower_bound(data.begin(), data.end(), val);
        if (*it == val)
            return; //去重
        data.insert(it, val);
    }

    vector<vector<int>> getIntervals() {
        vector<vector<int>> ret;
        if (data.empty())
            return ret;
        int size = data.size();
        int b = 0;
        for (int i = 1; i < size;++i) {
            if (data[i] != data[i - 1] + 1) {
                ret.push_back(vector<int>{data[b], data[i - 1]});
                b = i;
            }
        }
        ret.push_back(vector<int>{data[b], data.back()});
        return ret;
    }
private:
    deque<int> data;    //有序数组
};

class SummaryRanges1 {
public:
    /** Initialize your data structure here. */
    SummaryRanges1() {
        ret.clear();
    }

    void addNum(int val) {
        for (int i = 0; i < ret.size();++i) {
            if (ret[i][1] < val - 1)
                continue;   //需要查找下一个区间
            if (ret[i][0] == val || ret[i][1] == val)
                return;
            if (ret[i][0] > val + 1) {
                ret.insert(ret.begin() + i, vector<int>{val, val});
                return;
            }
            if (ret[i][0] == val + 1) {
                ret[i][0] = val;
                return;
            }
            if (ret[i][1] == val - 1) {
                ret[i][1] = val;
                if (i + 1 >= ret.size())
                    return;
                if (ret[i + 1][0] == val + 1) {
                    //合并区间
                    ret[i][1] = ret[i + 1][1];
                    ret.erase(ret.begin() + i + 1);
                }
                return;
            }
            return;
        }
        //遍历完整个区间集都没找到合适的位置就插在最后
        ret.push_back(vector<int>{val, val});
    }

    vector<vector<int>> getIntervals() {
        return ret;
    }
private:
    vector<vector<int>> ret;
};

int main(int argc, char const* argv[]) {
    SummaryRanges1 cls;
    cls.addNum(1);
    auto& r1 = cls.getIntervals();
    cout << "[";
    for (auto& v : r1) {
        cout << "[";
        cout << v[0] << "," << v[1];
        cout << "],";
    }
    cout << "]" << endl;
    cls.addNum(3);
    auto& r2 = cls.getIntervals();
    cout << "[";
    for (auto& v : r2) {
        cout << "[";
        cout << v[0] << "," << v[1];
        cout << "],";
    }
    cout << "]" << endl;
    cls.addNum(7);
    auto& r3 = cls.getIntervals();
    cout << "[";
    for (auto& v : r3) {
        cout << "[";
        cout << v[0] << "," << v[1];
        cout << "],";
    }
    cout << "]" << endl;
    cls.addNum(2);
    auto& r4 = cls.getIntervals();
    cout << "[";
    for (auto& v : r4) {
        cout << "[";
        cout << v[0] << "," << v[1];
        cout << "],";
    }
    cout << "]" << endl;
    cls.addNum(6);
    auto& r5 = cls.getIntervals();
    cout << "[";
    for (auto& v : r5) {
        cout << "[";
        cout << v[0] << "," << v[1];
        cout << "],";
    }
    cout << "]" << endl;
    return 0;
}

File: go/300.500/352.DataStreamAsDisjointIntervals.go
package main

import (
	"fmt"
	"sort"
)

type SummaryRanges struct {
	data []int
}

/** Initialize your data structure here. */
func Constructor() SummaryRanges {
	return SummaryRanges{
		data: make([]int, 0),
	}
}

func (this *SummaryRanges) AddNum(val int) {
	this.data = append(this.data, val)
	sort.Ints(this.data)
}

func (this *SummaryRanges) GetIntervals() [][]int {
	ret := [][]int{}
	size := len(this.data)
	b := 0
	for i := 1; i < size; i++ {
		if this.data[i] == this.data[i-1] {
			continue
		}
		if this.data[i] != this.data[i-1]+1 {
			ret = append(ret, []int{this.data[b], this.data[i-1]})
			b = i
		}
	}
	ret = append(ret, []int{this.data[b], this.data[len(this.data)-1]})
	return ret
}

type SummaryRanges1 struct {
	data [][]int
}

/** Initialize your data structure here. */
func Constructor1() SummaryRanges1 {
	return SummaryRanges1{
		data: make([][]int, 0),
	}
}

func (this *SummaryRanges1) AddNum(val int) {
	for i, v := range this.data {
		if v[1] < val-1 {
			continue
		}
		if v[0] == val || v[1] == val {
			return
		}
		if v[0] > val+1 {
			tmp := this.data[i:]
			tmp1 := make([][]int, len(tmp))
			copy(tmp1, tmp)
			this.data = this.data[:i]
			this.data = append(this.data, []int{val, val})
			this.data = append(this.data, tmp1...)
			return
		}
		if v[0] == val+1 {
			this.data[i][0] = val
			return
		}
		if v[1] == val-1 {
			this.data[i][1] = val
			if i+1 == len(this.data) {
				return
			}
			if this.data[i+1][0] == val+1 {
				this.data[i][1] = this.data[i+1][1]
				tmp := this.data[i+2:]
				this.data = this.data[:i+1]
				this.data = append(this.data, tmp...)
				return
			}
		}
		return
	}
	this.data = append(this.data, []int{val, val})
}

func (this *SummaryRanges1) GetIntervals() [][]int {
	return this.data
}

func main() {
	sum := Constructor1()
	sum.AddNum(6)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(6)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(0)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(4)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(8)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(7)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(6)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(4)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(7)
	fmt.Println(sum.GetIntervals())
	sum.AddNum(5)
	fmt.Println(sum.GetIntervals())
}
